drivers/power/max17851/max17851.c

```diff
@@ -0,0 +1,760 @@
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->single = init_param->single;
+	descriptor->baud_rate = init_param->baud_rate;
+
+	ret = max17851_reg_update(descriptor, MAX17851_CONFIG_GEN4_REG,
+				  MAX17851_MS_MASK,
+				  no_os_field_prep(MAX17851_MS_MASK, init_param->op_mode));
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->op_mode = init_param->op_mode;
+
+	ret = max17851_set_alert_debounce(descriptor, init_param->dbnc);
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->dbnc = init_param->dbnc;
+
+	ret = max17851_reg_write(descriptor, MAX17851_CONFIG_SAFEMON0_REG,
+				 init_param->gpio_rec_dly_csec);
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->gpio_rec_dly_csec = init_param->gpio_rec_dly_csec;
+
+	ret = max17851_reg_write(descriptor, MAX17851_CONFIG_SAFEMON1_REG,
+				 init_param->contact_tmr_dly_4xmin);
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->contact_tmr_dly_4xmin = init_param->contact_tmr_dly_4xmin;
+
+	ret = max17851_set_safemon_dly(descriptor, init_param->safemon_dly);
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->safemon_dly = init_param->safemon_dly;
+
+	ret = max17851_set_comm(descriptor, MAX17851_RETRY_COUNT_2,
+				init_param->comm_to_delay);
+	if (ret)
+		goto free_gpio2;
+
+	/* Wake up BMS chips in daisy-chain. */
+	ret = max17851_sel_alrtpckt_time(descriptor, MAX17851_ALRTPCKT_INFINITE);
+	if (ret)
+		goto free_gpio2;
+
+	ret = max17851_config_gen2(descriptor, MAX17851_TX_QUEUE_SEL, false);
+	if (ret)
+		goto free_gpio2;
+
+	descriptor->comm_to_delay = init_param->comm_to_delay;
+
+	/* Checking for errors on RX and clearing if necessary*/
+	ret = max17851_reg_read(descriptor, MAX17851_ALERT_GEN_REG, &reg_val);
+	if (ret)
+		goto free_gpio2;
+
+	if (reg_val) {
+		ret = max17851_reg_write(descriptor, MAX17851_ALERT_GEN_REG,
+					 0x00);
+		if (ret)
+			goto free_gpio2;
+	}
+
+	ret = max17851_clear_buff(descriptor, true);
+	if (ret)
+		goto free_gpio2;
+
+	ret = max17851_clear_buff(descriptor, false);
+	if (ret)
+		goto free_gpio2;
+
+	*desc = descriptor;
+
+	return 0;
+
+free_gpio2:
+	no_os_gpio_remove(descriptor->gpio2_desc);
+free_gpio1:
+	no_os_gpio_remove(descriptor->gpio1_desc);
+free_spi:
+	no_os_spi_remove(descriptor->spi_desc);
+free_desc:
+	no_os_free(descriptor);
+
+	return ret;
+}
+
+/**
+ * @brief Deallocate any resources used at the initialization function.
+ * @param desc - MAX17851 device descriptor.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int max17851_remove(struct max17851_desc *desc)
+{
+	no_os_gpio_remove(desc->gpio2_desc);
+	no_os_gpio_remove(desc->gpio1_desc);
+	no_os_spi_remove(desc->spi_desc);
+	no_os_free(desc);
+
+	return 0;
+}
+
+/**
+ * @brief MAX17851 platform specific UART platform ops structure
+ */
+const struct no_os_uart_platform_ops max17851_uart_ops = {
+	.init = &max17851_uart_init,
+	.read = &max17851_uart_read,
+	.write = &max17851_uart_write,
+	.get_errors = &max17851_uart_get_errors,
+	.remove = &max17851_uart_remove
+};
```

Comment:
Can we also implement the asynchronous functions? (or at least the `asynchronous_rx` functionality)